### ISSUE TITLE
Proposal: Dealing with null values in inverted index [accepted]

### ISSUE THREAD

**Issue by etiennedi** [semi-technologies/weaviate#2043]
## Tl;dr

The support around `null` or not-set values in Weaviate's objects is limited. While there is nothing to enforce that they are set, it's difficult to query the `null` state. This proposal aims to solve this.

## Stage 1 - Importing Null values:

During the discussions prior to this design, an interesting topic came up: What is the difference between a property that is not set and explicit `null`? Are explicit `null` values valid? 

### Proposal 
A property left out is the same as a property explicitly set to `null`. In other words, importing the following two objects into Weaviate leads to the same behavior:

```js
// object 1
{
  "prop1": "this one is set in both",
  "prop2": null
}

// object 2
{
  "prop1": "this one is set in both",
}
```

Notes:
* We have not yet validated if there are any technical blockers to this proposal, e.g from `go-openapi`, etc. The assumption is that this is feasible from a tech perspective.

## Stage 2 - Indexing null values
Indexing null values is an optional feature, as not every user needs this feature, and it takes up a non-negligible amount of space, as well as has a slight cost to import. 

### Proposal

The schema config is extended with two new configuration options which both default to `false`. They sit alongside the existing `"indexInverted": true`. The new options (combined with the existing option) looks like this:

```js
  "indexInverted": true, // defaults to true
  "indexNull": true, // defaults to false
  "indexLength": true // defaults to false
```

Some notes:
* The two new properties are only valid if `indexInverted` is also set to true. Should the user set this to `false`, there would be no inverted index at all, hence setting one of the two new props to `true` would error
* `indexNull` allows to query for null values, see "Stage 3"
* `indexLength` allows to query for prop (or in the case of arrayProp - array) length, see "Stage 4"

### Proposal
A new operator is introduced that matches all the objects where a property is either null or not null. As defined in the previous section "not set at all" and null are identical:

```js
// matching all null values
where: {
  operator: IsNull
  valueBoolean: true
  path: ["prop2"]
}

// matching all non-null values
where: {
  operator: IsNull
  valueBoolean: false
  path: ["prop2"]
}
```

Some notes:
* The poll on the Weaviate internal slack, as well as the discussion on the community slack both yielded `IsNull` as the favorite. The runner-up was `Exists`.
* While this proposal requires introducing a new operation `IsNull`, the value can only be `true` or `false`, thus the existing `valueBoolean` can be reused.
* Combining `operator: IsNull` with any other `value<Type>` field leads to a validation error

## Stage 4 - Filtering for an array or property length

The discussion above yielded another question about array length. Some argued that a zero-length array should be identical to a `null` property, others argued that they would explicitly distuingish between a `[]` and `null` array. As part of this, the desire to check for a specific length (other than 0) also came up. 

### Proposal

A new special syntax is introduced to specify that the length of the property is measured, not the property itself. This is mainly to distinguish between measuring the length vs. asserting on an individual value of the array. (Example, "less than 3" could mean an array with length less than 3 or it could mean an array of integers of which a value is less than three).

```js
where: {
  operator: LessThan
  valueInt: 17
  path: ["len(prop3)"]
}
```

Notes: 
* Whether this special syntax is feasible from a tech perspective, has not been validated yet.
* Length can be combined with any of the numerical operators, such as `Equal`, `LessThanEqual`, etc.
* Length is only defined for properties with variable length. Properties with fixed length, e.g. `int` or `number` which always take up exactly 8 bytes do not have a length property defined.
  * Measuring the length of an array returns the number of items in this array. This is consistent with most programming languages.
  * Measuring the length of a string/text prop returns the length of this particular string. This is consistent with most programming languages.
  * The length of an null-array is defined as 0. This is consistent with some programming languages, such as Golang. Note: You can still distuingish between a null-array and a zero-length array using the `IsNull` operator if you need to.

**Comment by etiennedi** [comment]
Not stale, part of v1.16 :-)

**Comment by etiennedi** [comment]
Released in `v1.16`.
